src/mscorlib/src/System/Array.cs

```diff
@@ -1559,31 +1559,23 @@ public static void Reverse<T>(T[] array)
         {
             if (array == null)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
-            Reverse(array, 0, array.Length);
+
+            // The Span ctor might perform an unnecessary type check if T is a reference type,
+            // so we use MemoryMarshal.CreateSpan to suppress this check.
+
+            MemoryMarshal.CreateSpan(ref Unsafe.As<byte, T>(ref array.GetRawSzArrayData()), array.Length).Reverse();
         }
 
         public static void Reverse<T>(T[] array, int index, int length)
         {
             if (array == null)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
-            if (index < 0)
-                ThrowHelper.ThrowIndexArgumentOutOfRange_NeedNonNegNumException();
-            if (length < 0)
-                ThrowHelper.ThrowLengthArgumentOutOfRange_ArgumentOutOfRange_NeedNonNegNum();
-            if (array.Length - index < length)
+            if ((uint)index > (uint)array.Length)
+                ThrowHelper.ThrowArgumentOutOfRange_IndexException();
+            if ((uint)length > (uint)(array.Length - index))
                 ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_InvalidOffLen);
 
-            ref T p = ref Unsafe.As<byte, T>(ref array.GetRawSzArrayData());
-            int i = index;
-            int j = index + length - 1;
-            while (i < j)
-            {
-                T temp = Unsafe.Add(ref p, i);
-                Unsafe.Add(ref p, i) = Unsafe.Add(ref p, j);
-                Unsafe.Add(ref p, j) = temp;
-                i++;
-                j--;
-            }
+            MemoryMarshal.CreateSpan(ref Unsafe.Add(ref Unsafe.As<byte, T>(ref array.GetRawSzArrayData()), index), length).Reverse();
```

Comment:
It would be useful to measure how much slower is `Array.Reverse` going to for small arrays.